### PR TITLE
Issue/backend

### PR DESCRIPTION
**Pull request by BelohvostovAlex**

### CODE REVIEW

File: src/components/TodoSelect/TodoSelect.tsx [hunk 1]
```diff
@@ -1,6 +1,12 @@
 import React from 'react';
 import { TodoSelectProps } from './interfaces';
 
+const availiableOptionsObj: Record<string, string> = {
+  todo: 'Todo',
+  in_progress: 'In progress',
+  done: 'Done',
+};
```
Comment (igorTT) [inline src/components/TodoSelect/TodoSelect.tsx]: create enum and use it throughout the whole application

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 import { TodoPage } from './TodoPage';
 import { ITodo, IPureTodo } from '../../models/ITodo';
-import { v4 } from 'uuid';
 
-const availiableOptions = ['Todo', 'In progress', 'Done'];
+const todoOptionsBackend: Record<string, string> = {
+  Todo: 'todo',
+  'In progress': 'in_progress',
+  Done: 'done',
+};
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: use enum

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 3]
```diff
@@ -21,14 +27,28 @@ export const TodoPageContainer: React.FC = () => {
...
   const [initialValue, setInitialValue] = useState(defaultValue as ITodo);
 
+  const fetchData = async (url: string) => {
+    const { data } = await axios.get<ITodo[]>(url);
+    setTodos(data);
+  };
+
+  useEffect(() => {
+    fetchData('http://localhost:4000/api');
+  }, []);
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: Create a custom hook- useFetchData

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 4]
```diff
@@ -21,14 +27,28 @@ export const TodoPageContainer: React.FC = () => {
...
+
   const hasTodo = !!filteredTodos.length;
 
-  const addTodo = (todo: IPureTodo) => {
-    setTodos([...todos, { ...todo, id: v4() }]);
+  const addTodo = async (todo: IPureTodo) => {
+    const { data } = await axios.post('http://localhost:4000/api/', todo);
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: Create a helper. Store it in a separate file.

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 5]
```diff
@@ -21,14 +27,28 @@ export const TodoPageContainer: React.FC = () => {
...
+    setTodos([...todos, data]);
     handleVisibleModal();
   };
 
-  const deleteTodo = (id: string) => {
+  const deleteTodo = async (id: string) => {
+    const currentTodo = todos.find((todo) => todo.id === id);
+    await axios.delete<ITodo>('http://localhost:4000/api/', {
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: create a helper.

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 6]
```diff
@@ -21,14 +27,28 @@ export const TodoPageContainer: React.FC = () => {
...
+      data: { ...currentTodo },
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: I don't see the reason for copying data

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 7]
```diff
@@ -52,18 +72,25 @@ export const TodoPageContainer: React.FC = () => {
...
-  const handleTodoProgress = (id: string, status: string) => {
+  const handleTodoProgress = async (id: string, status: string) => {
     setTodos(
       todos.map((todo) => (todo.id === id ? { ...todo, status: status } : todo))
     );
+    const currentTodo = todos.find((todo) => todo.id === id);
+    await axios.patch('http://localhost:4000/api/', {
+      ...currentTodo,
+      status: status,
+    });
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: create a helper

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 8]
```diff
@@ -72,13 +99,16 @@ export const TodoPageContainer: React.FC = () => {
...
+  const editTodo = async (todo: IPureTodo) => {
     const { id } = initialValue;
     const currTodo = todos.find((todo) => todo.id === id);
     const editedTodo = {
       ...currTodo,
       ...todo,
     };
+    await axios.patch('http://localhost:4000/api/', {
+      ...editedTodo,
+    });
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: the same as above

**Review by igorTT — CHANGES_REQUESTED** [review]

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 3 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: done

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 4 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: done

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 5 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: done

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 6 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: done

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 8 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: changed

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 7 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: done

File: package.json [hunk 9]
```diff
@@ -10,7 +10,9 @@
     "@types/node": "^16.11.36",
     "@types/react": "^18.0.9",
     "@types/react-dom": "^18.0.4",
+    "axios": "^0.27.2",
     "classnames": "^2.3.1",
+    "dotenv": "^16.0.1",
```
Comment (igorTT) [inline package.json]: I don't see it in use.

File: src/components/TodoSelect/TodoSelect.tsx [hunk 10]
```diff
@@ -1,5 +1,13 @@
 import React from 'react';
 import { TodoSelectProps } from './interfaces';
+import { availiableOptionsEnum } from '../../models/enums';
+
+const availiableOptionsObj: Record<string, string> = {
+  todo: availiableOptionsEnum.todo,
+  in_progress: availiableOptionsEnum.in_progress,
+  done: availiableOptionsEnum.done,
+};
+console.log(availiableOptionsObj);
```
Comment (igorTT) [inline src/components/TodoSelect/TodoSelect.tsx]: delete console.log

File: src/services/WebService.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    });
+    return data;
+  }
+
+  async updateData(value: ITodo) {
+    await axios.patch(REACT_APP_SERVER_URL!, value);
+  }
+}
+
+export default new WebService();
```
Comment (igorTT) [inline src/services/WebService.tsx]: Create an instance of the class only where you invoke it.

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 12]
```diff
@@ -2,9 +2,18 @@ import React, { useState, useEffect, useCallback } from 'react';
 
 import { TodoPage } from './TodoPage';
 import { ITodo, IPureTodo } from '../../models/ITodo';
-import { v4 } from 'uuid';
+import { useFetchData } from '../../hooks/useFetchData';
+import WebService from '../../services/WebService';
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx:6]: It's an instance of the class. It should be declared in a lower case

File: src/hooks/useFetchData.tsx [hunk 13]
```diff
@@ -0,0 +1,35 @@
+import axios from 'axios';
+import { useEffect, useState } from 'react';
+
+export const useFetchData = (url: string) => {
+  const [data, setData] = useState([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [hasError, setHasError] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  useEffect(() => {
```
Comment (igorTT) [inline src/hooks/useFetchData.tsx:11]: Add empty lines

File: src/hooks/useFetchData.tsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        if (response.status === 200) {
+          setData(response.data);
+          setIsLoading(false);
+          setHasError(false);
+          setErrorMessage('');
+          console.log(response);
+        } else {
+          setHasError(true);
+          setErrorMessage('Cant fetch the data');
+        }
```
Comment (igorTT) [inline src/hooks/useFetchData.tsx]: You'll get 100% percent 200 status. In other cases, for the GET request, you'll get only an error.

File: src/hooks/useFetchData.tsx [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    const fetchData = async () => {
+      try {
+        setIsLoading(true);
+        const response = await axios.get(url);
+        if (response.status === 200) {
+          setData(response.data);
+          setIsLoading(false);
+          setHasError(false);
+          setErrorMessage('');
+          console.log(response);
```
Comment (igorTT) [inline src/hooks/useFetchData.tsx]: Delete console.log

File: src/hooks/useFetchData.tsx (hunk 14 above)
Comment (igorTT) [inline src/hooks/useFetchData.tsx]: Add 'finally' block and set false value for loading.

File: src/hooks/useFetchData.tsx [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        setHasError(true);
+        setErrorMessage('Cant fetch the data');
+        setIsLoading(false);
+      }
+    };
+
+    fetchData();
+  }, [url]);
+
+  return { data, isLoading, hasError, errorMessage };
```
Comment (igorTT) [inline src/hooks/useFetchData.tsx]: You can replace hasError with hasError: !!errorMesasge

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 17]
```diff
@@ -2,9 +2,18 @@ import React, { useState, useEffect, useCallback } from 'react';
...
+import { availiableOptionsEnum } from '../../models/enums';
+
+const todoOptionsBackend: Record<string, string> = {
+  [availiableOptionsEnum.todo]: 'todo',
+  [availiableOptionsEnum.in_progress]: 'in_progress',
+  [availiableOptionsEnum.done]: 'done',
+};
 
-const availiableOptions = ['Todo', 'In progress', 'Done'];
+const availiableOptions = ['todo', 'in_progress', 'done'];
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx]: Why do we need to pass this property through the 5 components?

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 18]
```diff
@@ -20,15 +29,23 @@ export const TodoPageContainer: React.FC = () => {
...
-  const addTodo = (todo: IPureTodo) => {
-    setTodos([...todos, { ...todo, id: v4() }]);
+  const addTodo = async (todo: IPureTodo) => {
+    const data = await WebService.postData(todo);
+    setTodos([...todos, data]);
     handleVisibleModal();
   };
 
-  const deleteTodo = (id: string) => {
+  const deleteTodo = async (id: string) => {
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx:46]: I don't see the `await`

File: src/pages/TodoPage/TodoPageContainer.tsx [hunk 19]
```diff
@@ -52,18 +69,22 @@ export const TodoPageContainer: React.FC = () => {
     setInitialValue(defaultValue);
   };
 
-  const handleTodoProgress = (id: string, status: string) => {
+  const handleTodoProgress = async (id: string, status: string) => {
```
Comment (igorTT) [inline src/pages/TodoPage/TodoPageContainer.tsx:72]: I don't see the await

File: src/services/WebService.tsx [hunk 20]
```diff
@@ -0,0 +1,29 @@
+import axios from 'axios';
+import { IPureTodo, ITodo } from '../models/ITodo';
+
+const { REACT_APP_SERVER_URL } = process.env;
+
+class WebService {
+  async getData() {
```
Comment (igorTT) [inline src/services/WebService.tsx]: This method is not in use

**Review by igorTT — CHANGES_REQUESTED** [review]

File: src/components/TodoSelect/TodoSelect.tsx (hunk 10 above)
Comment (BelohvostovAlex) [reply src/components/TodoSelect/TodoSelect.tsx]: deleted

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 12 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx:6]: done

File: src/services/WebService.tsx (hunk 11 above)
Comment (BelohvostovAlex) [reply src/services/WebService.tsx]: done

File: src/hooks/useFetchData.tsx (hunk 13 above)
Comment (BelohvostovAlex) [reply src/hooks/useFetchData.tsx:11]: added

File: src/hooks/useFetchData.tsx (hunk 15 above)
Comment (BelohvostovAlex) [reply src/hooks/useFetchData.tsx]: deleted

File: src/hooks/useFetchData.tsx (hunk 14 above)
Comment (BelohvostovAlex) [reply src/hooks/useFetchData.tsx]: added, done

Comment (BelohvostovAlex) [reply src/hooks/useFetchData.tsx]: changed

File: src/hooks/useFetchData.tsx (hunk 16 above)
Comment (BelohvostovAlex) [reply src/hooks/useFetchData.tsx]: done

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 18 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx:46]: done

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 19 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx:72]: added

File: src/services/WebService.tsx (hunk 20 above)
Comment (BelohvostovAlex) [reply src/services/WebService.tsx]: deleted for this issue

File: package.json (hunk 9 above)
Comment (BelohvostovAlex) [reply package.json]: thought, had to install it to use and create .env file

File: src/pages/TodoPage/TodoPageContainer.tsx (hunk 17 above)
Comment (BelohvostovAlex) [reply src/pages/TodoPage/TodoPageContainer.tsx]: removed

**Review by igorTT — APPROVED** [review]

**Pull request merged**